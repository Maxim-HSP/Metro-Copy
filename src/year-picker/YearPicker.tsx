import React, { useState } from 'react';
import { DatePicker } from 'antd';
import * as moment from 'moment';
import { DatePickerProps, DatePickerMode } from 'antd/lib/date-picker/interface';

import './style/index.less';

export interface IYearPickerProps extends DatePickerProps {
  onChange?: (date: moment.Moment) => void;
  [props: string]: any;
}

const useYearPicker = (props: IYearPickerProps): IYearPickerProps => {
  const { onChange, ...restProps } = props;
  const [open, setOpen] = useState<boolean>(false);
  const [value, setValue] = useState<moment.Moment>(restProps.value || restProps.defaultValue);
  return {
    open,
    value,
    mode: 'year',
    showToday: false,
    format: 'YYYY',
    // 不使用onBlur或onFocus，会造成闪烁，因为每次点击都会触发这两个句柄
    onOpenChange: (open: boolean) => {
      if (open) setOpen(true)
      else setOpen(false)
    },
    onPanelChange: (val: moment.Moment, curMode: DatePickerMode) => {
      // 兼容value受控的情况
      if (!restProps.value) {
        setValue(val)
      }
      if (!curMode) {
        setOpen(false)
        if (onChange) onChange(val)
      }
    },
    // 兼容clear按钮，由于value是由onPanelChange维护的，故DatePicker的onChange只会在点击clear按钮时触发
    onChange: (val) => {
      if (!val) {
        setValue(val)
        if (onChange) onChange(val)
      }
    },
    ...restProps,
  };
};

const YearPicker: React.SFC<IYearPickerProps> = props => {
  return <DatePicker {...useYearPicker(props)} />;
};

// 兼容被Form注册为表单域的情况（form在注册表单域时会使用ref）
class ClassYearPicker extends React.PureComponent<IYearPickerProps, unknown> {
  render() {
    return <YearPicker {...this.props} />
  }
}

export {
  ClassYearPicker as default,
  YearPicker as PureYearPicker
}