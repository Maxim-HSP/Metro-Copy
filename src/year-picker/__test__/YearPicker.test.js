import React from 'react';
import { mount } from 'enzyme';
import YearPicker from '..';

describe('YearPicker',() => {
  let wrapper;
  beforeAll(() => {
    wrapper = mount(<YearPicker />);
  })

  it('render correctly', () => {
    expect(wrapper.render()).toMatchSnapshot();
    expect(wrapper.find('.ant-calendar-picker').length).toBe(1);
  });

  it('open panel', () => {
    wrapper.find('.ant-calendar-picker-input').simulate('focus');
    expect(wrapper.find('.ant-calendar-panel')).toHaveLength(1);
  });
  
  it('select year', () => {
    wrapper.find('.ant-calendar-picker-input').simulate('focus');
    expect(wrapper.find('.ant-calendar-panel')).toHaveLength(1);
    wrapper.find({title: '2018', role: 'gridcell'}).simulate('click');
    expect(wrapper.find('.ant-calendar-picker-input').prop('value')).toEqual('2018');
  })
});
