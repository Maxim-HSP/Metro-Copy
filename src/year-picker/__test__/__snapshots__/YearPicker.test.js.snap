// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YearPicker render correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "ant-calendar-picker",
    },
    "children": Array [
      Object {
        "attribs": Object {},
        "children": Array [
          Object {
            "attribs": Object {
              "class": "ant-calendar-picker-input ant-input",
              "placeholder": "Select date",
              "readonly": "",
              "value": "",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "aria-label": "icon: calendar",
                "class": "anticon anticon-calendar ant-calendar-picker-icon",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "aria-hidden": "true",
                    "class": "",
                    "data-icon": "calendar",
                    "fill": "currentColor",
                    "focusable": "false",
                    "height": "1em",
                    "viewBox": "64 64 896 896",
                    "width": "1em",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "d": "M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z",
                      },
                      "children": Array [],
                      "name": "path",
                      "namespace": "http://www.w3.org/2000/svg",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "d": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "d": undefined,
                      },
                    },
                  ],
                  "name": "svg",
                  "namespace": "http://www.w3.org/2000/svg",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "aria-hidden": undefined,
                    "class": undefined,
                    "data-icon": undefined,
                    "fill": undefined,
                    "focusable": undefined,
                    "height": undefined,
                    "viewBox": undefined,
                    "width": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "aria-hidden": undefined,
                    "class": undefined,
                    "data-icon": undefined,
                    "fill": undefined,
                    "focusable": undefined,
                    "height": undefined,
                    "viewBox": undefined,
                    "width": undefined,
                  },
                },
              ],
              "name": "i",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "aria-label": undefined,
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "aria-label": undefined,
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "placeholder": undefined,
              "readonly": undefined,
              "value": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "placeholder": undefined,
              "readonly": undefined,
              "value": undefined,
            },
          },
          Object {
            "attribs": Object {
              "aria-label": "icon: calendar",
              "class": "anticon anticon-calendar ant-calendar-picker-icon",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "aria-hidden": "true",
                  "class": "",
                  "data-icon": "calendar",
                  "fill": "currentColor",
                  "focusable": "false",
                  "height": "1em",
                  "viewBox": "64 64 896 896",
                  "width": "1em",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "d": "M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z",
                    },
                    "children": Array [],
                    "name": "path",
                    "namespace": "http://www.w3.org/2000/svg",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "d": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "d": undefined,
                    },
                  },
                ],
                "name": "svg",
                "namespace": "http://www.w3.org/2000/svg",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "aria-hidden": undefined,
                  "class": undefined,
                  "data-icon": undefined,
                  "fill": undefined,
                  "focusable": undefined,
                  "height": undefined,
                  "viewBox": undefined,
                  "width": undefined,
                },
                "x-attribsPrefix": Object {
                  "aria-hidden": undefined,
                  "class": undefined,
                  "data-icon": undefined,
                  "fill": undefined,
                  "focusable": undefined,
                  "height": undefined,
                  "viewBox": undefined,
                  "width": undefined,
                },
              },
            ],
            "name": "i",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "ant-calendar-picker-input ant-input",
                "placeholder": "Select date",
                "readonly": "",
                "value": "",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "placeholder": undefined,
                "readonly": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "placeholder": undefined,
                "readonly": undefined,
                "value": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "aria-label": undefined,
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "aria-label": undefined,
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
    ],
    "name": "span",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
